plutus-contract/src/Plutus/Contract/Test/StateModel.hs

```diff
@@ -0,0 +1,176 @@
+
+instance Eq (Any f) where
+  Some (a :: f a) == Some (b :: f b) =
+    case eqT @a @b of
+      Just Refl -> a == b
+      Nothing   -> False
+  Error s == Error s' = s == s'
+  _ == _ = False
+
+data Step state where
+  (:=) :: (Show a, Typeable a, Eq (Action state a), Typeable (Action state a), Show (Action state a)) =>
+            Var a -> Action state a -> Step state
+
+infix 5 :=
+
+deriving instance (forall a. Show (Action state a)) => Show (Step state)
+
+newtype Var a = Var Int
+  deriving (Eq, Ord, Show, Typeable)
+
+instance Eq (Step state) where
+  (Var i := act) == (Var j := act') =
+    (i==j) && Some act == Some act'
+
+newtype Actions state = Actions [Step state]
+  deriving Eq
+
+instance (forall a. Show (Action state a)) => Show (Actions state) where
+  showsPrec d (Actions as)
+    | d>10      = ("("++).showsPrec 0 (Actions as).(")"++)
+    | null as   = ("Actions []"++)
+    | otherwise = (("Actions \n [")++) .
+                  foldr (.) (showsPrec 0 (last as) . ("]"++))
+                    [showsPrec 0 a . (",\n  "++) | a <- init as]
+
+
+instance (Typeable state, StateModel state) => Arbitrary (Actions state) where
+  arbitrary = Actions <$> arbActions initialState 1
+    where
+      arbActions :: state -> Int -> Gen [Step state]
+      arbActions s step = sized $ \n ->
+        let w = n `div` 2 + 1 in
+          frequency [(1, return []),
+                     (w, do mact <- arbitraryAction s `suchThatMaybe`
+                                      \a -> case a of
+                                              Some act -> precondition s act
+                                              Error _  -> False
+                            case mact of
+                              Just (Some act) ->
+                                ((Var step := act):) <$> arbActions (nextState s act (Var step)) (step+1)
+                              Just Error{} -> error "impossible"
+                              Nothing ->
+                                return [])]
+
+  shrink (Actions as) =
+    map (Actions . prune . map fst) (shrinkList shrinker (withStates as))
+    where shrinker ((Var i := act),s) = [((Var i := act'),s) | Some act' <- shrinkAction s act]
+
+prune :: StateModel state => [Step state] -> [Step state]
+prune = loop initialState
+  where loop _s [] = []
+        loop s ((var := act):as)
+          | precondition s act
+            = (var := act):loop (nextState s act var) as
+          | otherwise
+            = loop s as
+
+
+withStates :: StateModel state => [Step state] -> [(Step state,state)]
+withStates = loop initialState
+  where
+    loop _s [] = []
+    loop s ((var := act):as) =
+      ((var := act),s):loop (nextState s act var) as
+
+stateAfter :: StateModel state => Actions state -> state
+stateAfter (Actions script) = loop initialState script
+  where
+    loop s []                  = s
+    loop s ((var := act) : as) = loop (nextState s act var) as
+
+runScript :: StateModel state =>
```

Comment:
Script -> Actions (probably worth just doing a global search for Script to be sure)